pkg/apis/core/types.go

```diff
@@ -4755,3 +4761,42 @@ const (
 	// DefaultHardPodAffinityWeight defines the weight of the implicit PreferredDuringScheduling affinity rule.
 	DefaultHardPodAffinitySymmetricWeight int32 = 1
 )
+
+type UnsatisfiableConstraintResponse string
+
+const (
+	// DoNotSchedule denotes do not schedule a pod in all circumstances
+	DoNotSchedule UnsatisfiableConstraintResponse = "DoNotSchedule"
+	// ScheduleAnyway denotes schedule a pod despite of any circumstance
+	ScheduleAnyway UnsatisfiableConstraintResponse = "ScheduleAnyway"
+)
+
+// TopologySpreadConstraint is the topological spread constraint for a pod.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree of imbalance of pods distribution.
+	// It's the max difference between the number of matching pods in any two
+	// topology domains against a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// spread as 1/1/0:
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required value. Default value is 1 and 0 is not allowed.
+	MaxSkew int32
+	// TopologyKey is the key such that we consider each value as a "bucket";
+	// we try to put balanced number of pods into each bucket.
+	// It's a required field.
+	TopologyKey string
+	// WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+	// the spread constraint.
+	// - DoNotSchedule (default) tells the scheduler not to schedule it
+	// - ScheduleAnyway tells the scheduler to still schedule it
+	// Note: it's considered as "Unsatisfiable" only when actual skew on all nodes
+	// exceeds "MaxSkew".
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintResponse
+	// Label selector for pods. This's enforced by scheduler to check which pods
```

Comment:
// LabelSelector is used to find matching pods. Pods that match this label selector are counted to determine the number of pods in their corresponding topology domain.